python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -608,6 +608,151 @@ def _scan(self, op: str, *args, **kwargs):
 
     aggregate = agg
 
+    def _head_tail(self, n, *, take_head=True):
+        """Return the head or tail of each group
+
+        Parameters
+        ----------
+        n
+           Number of entries to include (if negative, number of
+           entries to exclude)
+        take_head
+           Do we want the head or the tail of the group
+
+        Returns
+        -------
+        New DataFrame or Series
+
+        Notes
+        -----
+        Unlike pandas, this returns an object in group order, not
+        original order
```

Comment:
Is it possible to optionally preserve order without introducing a large performance hit? For scenarios in which behavioral alignment with pandas is desired.

Reply:
I think so, let me check...

Reply:
I added a preserve_order keyword argument. It costs a bit.
```
df = create_df(
    100_000_000, # num rows
    1_000_000, # num groups
)

In [7]: %timeit df.groupby("a").head(2, preserve_order=False)
207 ms ± 1.18 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [8]: %timeit df.groupby("a").head(2, preserve_order=True)
269 ms ± 742 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
I'm inclined to say that this is a case where users must opt _in_ to performance rather than opt _out_. That is, I'm proposing that `preserve_order=True` be the default.

I'd be more comfortable with opt out if `preserve_order=` was a kwarg that Pandas supported and we used a different default. But here we are introducing a new kwarg and making the default behaviour different from Pandas. This makes it impossible for someone to write their code to work identically with Pandas and cuDF.

Thoughts?

Reply:
Yes, but the `sort=` keyword is one that Pandas supports for merge/groupby. So one could write their code to work identically with cuDF and Pandas by explicitly passing `sort=True` everywhere. That is not possible here. 

I agree that when we do support a single config option to control sorting behaviour across all APIs, we can change the default behaviour here to be `preserve_order=False`.

Reply:
I think this is a difficult circle to square. My feeling is that the result order for many pandas operations being consistent with the input is merely an implementation detail and not part of the API contract. I am not sure how many people are relying on this: one would hope not many since any perturbation of their input data could (potentially) change output results.

Reply:
In the context of aggregations, `sort=True` will sort the result by the group keys. Since there is only one row per group key in the result of an aggregation, cuDF and Pandas should return the same result when `sort=True`, regardless of whether a hash or sort-based  groupby was used internally.

Reply:
Unfortunately, it appears that it is a guarantee. From https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html:

> Groupby preserves the order of rows within each group.

Reply:
> Groupby preserves the order of rows within each group.

This we are doing! 

The difference in ordering for the `head`/`tail` implementation is the pandas implementation does:

"Figure out which rows in the input dataframe correspond to the first `n` rows of each group, select those rows"

Whereas my implementation does "get the grouped dataframe, select the first `n` rows of each group". So it's a global ordering difference.

So if your dataframe is:
```
  a b
0 1 0
1 0 1
2 1 2
3 3 3
4 0 4
5 3 5
```
Then pandas `groupby("a").head(2)` produces
```
  a b
0 1 0
1 0 1
2 1 2
3 3 3
4 0 4
5 3 5
```

Whereas my implementation (without `preserve_order=True`) produces
```
   a b
1 0 1
4 0 4
0 1 0
2 1 2
3 3 3
5 3 5
```

So the order in each group is preserved

Reply:
I guess I'm a bit confused now. What ordering (if any) are we unsure about being an implementation detail of Pandas and versus an API guarantee?

Reply:
Sorry, pandas [groupby.head](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.GroupBy.head.html) says:

> Return first n rows of each group.

> Similar to .apply(lambda x: x.head(n)), but it returns a subset of rows from the original DataFrame with **original index and order preserved** (as_index flag is ignored).

(Emphasis mine).

So I guess that this does say "`head` really acts like subsampling and doesn't sort".

My implementation needs to undo a sort to get back to the original order of the dataframe. The question is, is whether two dataframes with the same index (just sorted differently) are considered equivalent enough.

Reply:
Gotcha - of course the answer to that depends on how people write their code. But generally speaking, I would consider them not equivalent, because when using the default (RangeIndex), one may expect to index with `iloc[]` rather than `loc[]`.

Reply:
> From what I remember seeing in a few old threads it was originally an implementation detail but something that too many users came to rely on for them to remove safely (@mroeschke could correct me on that).

Yeah the default sorting with groupby of pandas just came out of the original implementation. I don't recall seeing threads of users directly worried of this changing, but it's highly theorized by the core team that since it's been the default for a while there's some hesitation in changing it: https://github.com/pandas-dev/pandas/pull/48809

Reply:
OK, to resolve this discussion in this PR, I'll flip the default to match pandas, but leave the note about performance in the docstring.